### PR TITLE
Factor calculation of progress from chunkmap

### PR DESCRIPTION
**Pull request by mattw-mega**
Two nearly identical pieces of code, for figuring out our current upload/download progress
from the chunkmap records - just one extra line in one - can easily be a function.

### CODE REVIEW

File: src/transferslot.cpp [hunk 1]
```diff
@@ -353,6 +353,36 @@ bool chunkmac_map::unserialize(const char*& ptr, const char* end)
...
+        {
+            m_off_t chunksize = chunkceil - ChunkedHash::chunkfloor(it->first);
+            progresscompleted += chunksize;
+        }
+        else
+        {
+            progresscompleted += it->second.offset;
+            if (lastblockprogress)
+            {
+                *lastblockprogress = it->second.offset;
```
Comment (cb-mega) [inline src/transferslot.cpp]: I think this should be `+=`

File: src/transferslot.cpp [hunk 2]
```diff
@@ -353,6 +353,36 @@ bool chunkmac_map::unserialize(const char*& ptr, const char* end)
     return true;
 }
 
+void chunkmac_map::calcprogress(m_off_t size, m_off_t& chunkpos, m_off_t& progresscompleted, m_off_t* lastblockprogress)
+{
+    chunkpos = 0;
```
Comment (cb-mega) [inline src/transferslot.cpp]: I don't see where these two variables are set to 0 in transfer.cpp? Did the original code do that?

File: src/transferslot.cpp [hunk 3]
```diff
@@ -353,6 +353,36 @@ bool chunkmac_map::unserialize(const char*& ptr, const char* end)
...
+    progresscompleted = 0;
+
+    for (chunkmac_map::iterator it = begin(); it != end(); it++)
```
Comment (cb-mega) [inline src/transferslot.cpp]: I'd change this to `++it`

File: src/transferslot.cpp (hunk 1 above)
Comment (mattw-mega) [reply src/transferslot.cpp]: good point, updated - so it's exactly the same

File: src/transferslot.cpp (hunk 2 above)
Comment (mattw-mega) [reply src/transferslot.cpp]: In transfer.cpp, those are passed from the `t = new Transfer` object, and have been initialised to 0 in its constructor.  good to check though, thanks!

File: src/transferslot.cpp (hunk 3 above)
Comment (mattw-mega) [reply src/transferslot.cpp]: done :-)

File: src/transferslot.cpp (hunk 2 above)
Comment (cb-mega) [reply src/transferslot.cpp]: Isn't that a weird side-effect of the function to set these variables to 0 if the map is empty for instance?

Comment (cb-mega) [reply src/transferslot.cpp]: Maybe those should be set to 0 by the caller?

**Pull request merged**